java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -262,11 +244,28 @@ private RelationState implies(RelationalSymbolicValue relation) {
     return RelationState.UNDETERMINED;
   }
 
-  private List<RelationalSymbolicValue> transitiveRelations(ProgramState programState) {
-    return knownRelations(programState)
-      .map(this::deduceTransitiveOrSimplified)
-      .filter(Objects::nonNull)
-      .collect(Collectors.toList());
+  private Set<RelationalSymbolicValue> transitiveRelations(Set<RelationalSymbolicValue> knownRelations) {
+    Set<RelationalSymbolicValue> newRelations = new HashSet<>();
+    Deque<RelationalSymbolicValue> workList = new ArrayDeque<>();
+    int iterations = 0;
+    workList.add(this);
+    while (!workList.isEmpty()) {
+      if (newRelations.size() > MAX_DEDUCED_RELATIONS || iterations > MAX_ITERATIONS) {
+        // safety mechanism in case of an error in the algorithm
+        // should not happen under normal conditions
+        throw new RelationalSymbolicValue.TransitiveRelationExceededException("Used relations: " + newRelations.size() + ". Iterations " + iterations);
+      }
+      iterations++;
+      RelationalSymbolicValue relation = workList.pop();
+      for (RelationalSymbolicValue knownRelation : knownRelations) {
+        RelationalSymbolicValue r = relation.deduceTransitiveOrSimplified(knownRelation);
+        if (r != null && !knownRelations.contains(r) && !newRelations.contains(r)) {
+          newRelations.add(r);
```

Comment:
could be done slightly better with  : 
```
 if (r != null && !knownRelations.contains(r)) {
   if(newRelations.add(r)) {
      worklist.add(r);
   }
 }
```
and those two if can be in turn collapsed : Wonder about readability though...